atxm/machine.py

```diff
@@ -352,33 +331,39 @@ def __strategize(self) -> None:
 
         try:
             txhash = self.__fire(tx=_active_copy, msg=_names)
-        except InsufficientFunds:
-            # special case re-raise insufficient funds (for now)
-            # TODO #13
-            # TODO should the following also be done?
-            #  self._tx_tracker.update_failed_retry_attempt(_active_copy)
-            raise
-        except (ValidationError, Web3Exception, ValueError) as e:
-            self._tx_tracker.update_failed_retry_attempt(_active_copy)
+        except InsufficientFunds as e:
+            # special case
+            self.log.error(
+                f"[insufficient funds] transaction #atx-{_active_copy.id}|{_active_copy.params['nonce']} "
+                f"failed because of insufficient funds - {e}"
+            )
+            # get hook from actual pending object (not a deep copy)
+            hook = self._tx_tracker.pending.on_insufficient_funds
+            fire_hook(hook, _active_copy, e)
+            return
+        except Exception as e:
+            self._tx_tracker.update_active_after_failed_strategy_update(_active_copy)
             self.__handle_retry_failure(_active_copy, e)
             return
 
         _active_copy.txhash = txhash
-        self._tx_tracker.update_after_retry(_active_copy)
+        self._tx_tracker.update_active_after_successful_strategy_update(_active_copy)
 
         pending_tx = self._tx_tracker.pending
-        self.log.info(f"[retry] transaction #{pending_tx.id} has been re-broadcasted")
+        self.log.info(
+            f"[retry] transaction #{pending_tx.id} has been re-broadcasted with updated params"
+        )
         if pending_tx.on_broadcast:
             fire_hook(hook=pending_tx.on_broadcast, tx=pending_tx)
 
     def __handle_retry_failure(self, attempted_tx: PendingTx, e: Exception):
-        log.warn(
+        self.log.warn(
```

Comment:
should we do the same on lines 225 and 235?

Reply:
✔️